### PR TITLE
fix(pnp): handle private import mappings

### PR DESCRIPTION
**Pull request by merceyz**
**What's the problem this PR addresses?**

Node.js supports an `imports` field in `package.json` that the PnP API doesn't support yet.
https://nodejs.org/dist/latest-v18.x/docs/api/packages.html#imports

Fixes https://github.com/yarnpkg/berry/issues/3843
Ref https://github.com/nodejs/node/issues/40579

**How did you fix it?**

Extracted the required code from Node.js and adapted it to work with PnP.

**Checklist**
- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).
- [x] I have set the packages that need to be released for my changes to be effective.
- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

**Comment by RDIL** [comment]
For the Node core code, should there be a plan in place, for when and how to update it?

**Comment by arcanis** [comment]
> For the Node core code, should there be a plan in place for when and how to update it?

Usually we update whenever someone tells us something doesn't seem to work compared to Node - it's not ideal, and hopefully it'll work better if Node provides a blessed resolution library.

File: packages/yarnpkg-pnp/sources/loader/makeApi.ts [hunk 1]
```diff
@@ -863,6 +880,28 @@ export function makeApi(runtimeState: RuntimeState, opts: MakeApiOptions): PnpAp
...
+          return resolveUnqualified(npath.toPortablePath(fileURLToPath(resolved)), {extensions});
+        } else {
+          if (resolved.startsWith(`#`))
+            // Node behaves interestingly by default so just block the request for now.
+            // https://github.com/nodejs/node/issues/40579
+            throw new Error(`Mapping from one private import to another isn't allowed`);
+
+          return resolveRequest(resolved as PortablePath, issuer, {conditions, considerBuiltins});
+        }
+      }
```
Comment (arcanis) [inline packages/yarnpkg-pnp/sources/loader/makeApi.ts:903]: Can you move this in a dedicated function (`resolvePrivateRequest`?)? No need to export it, I think, but it'd at least improve the readability.

File: packages/yarnpkg-pnp/sources/loader/makeApi.ts [hunk 2]
```diff
@@ -1,9 +1,12 @@
 import {ppath, Filename}                                                                                                                                                                   from '@yarnpkg/fslib';
 import {FakeFS, NativePath, PortablePath, VirtualFS, npath}                                                                                                                                from '@yarnpkg/fslib';
+import assert                                                                                                                                                                              from 'assert';
```
Comment (arcanis) [inline packages/yarnpkg-pnp/sources/loader/makeApi.ts:3]: We don't use `assert` at the moment, it should be a sweeping change in another PR

**Review by arcanis — APPROVED** [review]
